File: SASGMACustomEventBanner.m
//
//  SASGMACustomEventBanner.m
//
//  Created by Julien Gomez on 21/06/16.
//  Copyright © 2016 Smart AdServer. All rights reserved.
//

#import "SASGMACustomEventBanner.h"
#import "SASBannerView.h"
#import "SASAdView+GMA.h"


@interface SASGMACustomEventBanner () <SASAdViewDelegate>

@property(nonatomic, strong) SASBannerView *bannerView;

@end

@implementation SASGMACustomEventBanner

@synthesize delegate;

#pragma mark GMACustomEventBanner implementation

- (void)requestBannerAd:(GADAdSize)adSize
              parameter:(NSString *)serverParameter
                  label:(NSString *)serverLabel
                request:(GADCustomEventRequest *)request {

    // Create the bannerView with the appropriate size.
    self.bannerView = [[SASBannerView alloc] initWithFrame:CGRectMake(0, 0, adSize.size.width, adSize.size.height)];
    self.bannerView.modalParentViewController = self.delegate.viewControllerForPresentingModalView;
    self.bannerView.delegate = self;

    [self.bannerView loadFormatWithDFPServerParameter:serverParameter request:request];

}

#pragma mark SASAdViewDelegate implementation

- (void)adViewDidLoad:(SASAdView *)adView {
    [self.delegate customEventBanner:self didReceiveAd:adView];
}


- (void)adView:(SASAdView *)adView didFailToLoadWithError:(NSError *)error {
    [self.delegate customEventBanner:self didFailAd:error];
}


- (BOOL)adView:(SASAdView *)adView shouldHandleURL:(NSURL *)URL {
    [self.delegate customEventBannerWasClicked:self];
    return YES;
}


- (void)adViewWillPresentModalView:(SASAdView *)adView {
    [self.delegate customEventBannerWillPresentModal:self];
}


- (void)adViewWillDismissModalView:(SASAdView *)adView {
    [self.delegate customEventBannerWillDismissModal:self];
}

@end

File: SASGMACustomEventBanner.swift
//
//  SASGMACustomEventBanner.swift
//  NewSample
//
//  Created by Guillaume Laubier on 13/09/2021.
//

import Foundation
import SASDisplayKit
import GoogleMobileAds

@objc(SASGMACustomEventBanner)
class SASGMACustomEventBanner : NSObject, GADCustomEventBanner, SASBannerViewDelegate {
    var delegate: GADCustomEventBannerDelegate?
    
    private var bannerView: SASBannerView?
    
    required override init() {
    }
    
    func requestAd(_ adSize: GADAdSize, parameter serverParameter: String?, label serverLabel: String?, request: GADCustomEventRequest) {
        guard let placement = SASGMAUtils.placementWith(serverParameter: serverParameter, request: request, extras: nil) else {
            // Placement is invalid, sending an error
            let error = NSError(domain: SASGMAUtils.kSASGMAErrorDomain, code: SASGMAUtils.kSASGMAErrorCodeInvalidServerParameters, userInfo: nil)
            self.delegate?.customEventBanner(self, didFailAd: error)
            return
        }
        
        // Create the bannerView with the appropriate size
        bannerView = SASBannerView(frame: CGRect(x: 0, y: 0, width: adSize.size.width, height: adSize.size.height))
        bannerView?.modalParentViewController = delegate?.viewControllerForPresentingModalView
        bannerView?.delegate = self
        
        // Load the previously retrieved ad placement
        bannerView?.load(with: placement)
    }
    
    // Pragma - SASBannerViewDelegate implementation
    
    func bannerViewDidLoad(_ bannerView: SASBannerView) {
        delegate?.customEventBanner(self, didReceiveAd: bannerView)
    }
    
    func bannerView(_ bannerView: SASBannerView, didFailToLoadWithError error: Error) {
        delegate?.customEventBanner(self, didFailAd: error)
    }
    
    func bannerView(_ bannerView: SASBannerView, didClickWith URL: URL) {
        delegate?.customEventBannerWasClicked(self)
    }
    
    func bannerViewWillPresentModalView(_ bannerView: SASBannerView) {
        delegate?.customEventBannerWillPresentModal(self)
    }
    
    func bannerViewWillDismissModalView(_ bannerView: SASBannerView) {
        delegate?.customEventBannerWillDismissModal(self)
    }
    
}
